File: src/tools/get-reservation-notes.ts
/**
 * Herramienta MCP para obtener notas de una reservación específica de Track HS
 */

import { BaseTrackHSTool } from '../core/base-tool.js';
import { ReservationNotesResponse, ReservationNotesParams } from '../types/reservation-notes.js';

export class GetReservationNotesTool extends BaseTrackHSTool {
  name = 'get_reservation_notes';
  description = 'Get notes for a specific reservation with optional filtering and pagination';
  
  inputSchema = {
    type: 'object' as const,
    properties: {
      reservationId: { 
        type: 'string', 
        description: 'The ID of the reservation to get notes for'
      },
      // Paginación
      page: { 
        type: 'number', 
        minimum: 0,
        description: 'Page number of result set'
      },
      size: { 
        type: 'number', 
        minimum: 1,
        maximum: 100,
        description: 'Size of page (max 100)'
      },
      // Filtros
      isInternal: { 
        type: 'boolean', 
        description: 'Filter by internal notes (true) or external notes (false)'
      },
      noteType: { 
        type: 'string', 
        description: 'Filter by note type'
      },
      priority: { 
        type: 'string', 
        enum: ['low', 'medium', 'high'],
        description: 'Filter by note priority'
      },
      author: { 
        type: 'string', 
        description: 'Filter by note author'
      },
      // Ordenamiento
      sortBy: { 
        type: 'string', 
        enum: ['createdAt', 'updatedAt', 'author', 'priority'],
        default: 'createdAt',
        description: 'Field to sort notes by'
      },
      sortDirection: { 
        type: 'string', 
        enum: ['asc', 'desc'],
        default: 'desc',
        description: 'Sort direction'
      },
      // Búsqueda
      search: { 
        type: 'string', 
        description: 'Search in note content'
      },
      dateFrom: { 
        type: 'string', 
        format: 'date-time',
        description: 'Filter notes from this date (ISO 8601 format)'
      },
      dateTo: { 
        type: 'string', 
        format: 'date-time',
        description: 'Filter notes to this date (ISO 8601 format)'
      }
    },
    required: ['reservationId']
  };

  async execute(params: ReservationNotesParams & { reservationId: string }): Promise<ReservationNotesResponse> {
    // Validar parámetros
    this.validateParams(params);

    const { reservationId, ...queryParams } = params;

    // Validar que el ID no esté vacío
    if (!reservationId || reservationId.trim() === '') {
      throw new Error('El ID de reservación no puede estar vacío');
    }

    // Construir query parameters
    const urlParams = new URLSearchParams();
    
    // Agregar parámetros de paginación
    if (queryParams.page !== undefined) {
      urlParams.append('page', queryParams.page.toString());
    }
    if (queryParams.size !== undefined) {
      urlParams.append('size', queryParams.size.toString());
    }

    // Agregar filtros
    if (queryParams.isInternal !== undefined) {
      urlParams.append('isInternal', queryParams.isInternal.toString());
    }
    if (queryParams.noteType) {
      urlParams.append('noteType', queryParams.noteType);
    }
    if (queryParams.priority) {
      urlParams.append('priority', queryParams.priority);
    }
    if (queryParams.author) {
      urlParams.append('author', queryParams.author);
    }

    // Agregar ordenamiento
    if (queryParams.sortBy) {
      urlParams.append('sortBy', queryParams.sortBy);
    }
    if (queryParams.sortDirection) {
      urlParams.append('sortDirection', queryParams.sortDirection);
    }

    // Agregar búsqueda
    if (queryParams.search) {
      urlParams.append('search', queryParams.search);
    }
    if (queryParams.dateFrom) {
      urlParams.append('dateFrom', queryParams.dateFrom);
    }
    if (queryParams.dateTo) {
      urlParams.append('dateTo', queryParams.dateTo);
    }

    const queryString = urlParams.toString();
    const endpoint = `/pms/reservations/${encodeURIComponent(reservationId)}/notes${queryString ? `?${queryString}` : ''}`;
    
    try {
      const result = await this.apiClient.get<ReservationNotesResponse>(endpoint);
      return result;
    } catch (error) {
      if (error instanceof Error && error.message.includes('404')) {
        throw new Error(`Reservación con ID '${reservationId}' no encontrada o no tiene notas`);
      }
      throw new Error(`Error al obtener notas de reservación: ${error instanceof Error ? error.message : 'Error desconocido'}`);
    }
  }
}

File: src/trackhs_mcp/tools/get_reservation_notes.py
"""
Herramienta MCP para obtener notas de reserva de Track HS API
"""

from typing import Optional
from ..core.api_client import TrackHSApiClient

def register_get_reservation_notes(mcp, api_client: TrackHSApiClient):
    """Registra la herramienta get_reservation_notes"""
    
    @mcp.tool()
    async def get_reservation_notes(
        reservation_id: int,
        page: int = 1,
        size: int = 10,
        is_internal: bool = None,
        note_type: str = None,
        priority: str = None,
        author: str = None,
        sort_by: str = "createdAt",
        sort_direction: str = "desc",
        search: str = None,
        date_from: str = None,
        date_to: str = None
    ):
        """
        Get reservation notes from Track HS
        
        Args:
            reservation_id: ID de la reserva
            page: Número de página (default: 1)
            size: Tamaño de página (default: 10)
            is_internal: Si es nota interna (opcional)
            note_type: Tipo de nota (opcional)
            priority: Prioridad de la nota (opcional)
            author: Autor de la nota (opcional)
            sort_by: Campo para ordenar (default: "createdAt")
            sort_direction: Dirección de ordenamiento (default: "desc")
            search: Término de búsqueda (opcional)
            date_from: Fecha desde (opcional)
            date_to: Fecha hasta (opcional)
        """
        query_params = {
            "page": page,
            "size": size,
            "sortBy": sort_by,
            "sortDirection": sort_direction
        }
        
        if is_internal is not None:
            query_params["isInternal"] = is_internal
        if note_type:
            query_params["noteType"] = note_type
        if priority:
            query_params["priority"] = priority
        if author:
            query_params["author"] = author
        if search:
            query_params["search"] = search
        if date_from:
            query_params["dateFrom"] = date_from
        if date_to:
            query_params["dateTo"] = date_to
        
        endpoint = f"/reservations/{reservation_id}/notes"
        query_string = "&".join([f"{k}={v}" for k, v in query_params.items()])
        if query_string:
            endpoint += f"?{query_string}"
        
        try:
            result = await api_client.get(endpoint)
            return result
        except Exception as e:
            return {"error": f"Error al obtener notas de reserva: {str(e)}"}
